python/ray/train/tests/test_results_preprocessors.py

```diff
@@ -39,6 +42,68 @@ def test_sequential_results_preprocessor():
     assert preprocessed_results == expected
 
 
+def test_average_results_preprocessor():
+    from copy import deepcopy
+    import numpy as np
+
+    results = [{"a": 1, "b": 2}, {"a": 3, "b": 4}, {"a": 5, "b": 6}, {"a": 7, "b": 8}]
+    expected = deepcopy(results)
+    for res in expected:
+        res.update(
+            {
+                "Average(a)": np.mean([result["a"] for result in results]),
+                "Average(b)": np.mean([result["b"] for result in results]),
+            }
+        )
+
+    preprocessor = AverageResultsPreprocessor(["a", "b"])
+    preprocessed_results = preprocessor.preprocess(results)
+
+    assert preprocessed_results == expected
+
+
+def test_max_results_preprocessor():
+    from copy import deepcopy
+    import numpy as np
+
+    results = [{"a": 1, "b": 2}, {"a": 3, "b": 4}, {"a": 5, "b": 6}, {"a": 7, "b": 8}]
+    expected = deepcopy(results)
+    for res in expected:
+        res.update(
+            {
+                "Max(a)": np.max([result["a"] for result in results]),
+                "Max(b)": np.max([result["b"] for result in results]),
+            }
+        )
+
+    preprocessor = MaxResultsPreprocessor(["a", "b"])
+    preprocessed_results = preprocessor.preprocess(results)
+
+    assert preprocessed_results == expected
+
+
+def test_weighted_average_results_preprocessor():
+    from copy import deepcopy
+    import numpy as np
+
+    results = [{"a": 1, "b": 2}, {"a": 3, "b": 4}, {"a": 5, "b": 6}, {"a": 7, "b": 8}]
+    expected = deepcopy(results)
+    total_weight = np.sum([result["b"] for result in results])
+    for res in expected:
+        res.update(
+            {
+                "Weighted average [by b](a)": np.sum(
+                    [result["a"] * result["b"] / total_weight for result in results]
+                )
+            }
+        )
+
+    preprocessor = WeightedAverageResultsPreprocessor(["a"], "b")
+    preprocessed_results = preprocessor.preprocess(results)
+
+    assert preprocessed_results == expected
+
+
```

Comment:
Can you also add tests for the degenerate cases?

What happens when some workers report the metric, but others don't?
What happens when the metric does not have a valid value type?
What happens when some workers report with a valid value type but others don't?
What happens when some workers report the weight key but others don't?